napari/_qt/qt_viewer.py

```diff
@@ -690,12 +690,17 @@ def _save_layers_dialog(self, selected=False):
         dlg = QFileDialog()
         hist = get_save_history()
         dlg.setHistory(hist)
-
+        # get the layer's name to use for a default name if only one layer is selected
+        selected_layer_name = ''
+        if self.viewer.layers.selection.active is not None:
+            selected_layer_name = self.viewer.layers.selection.active.name
```

Comment:
Layer names often include spaces, we don't want those in file names. For this to go in I think we should snake_case everything beforehand!

Reply:
not only snake case, but also remove every special characters like `/`, `\` etc

Reply:
A quick and dirty way that I *think* works is to just use `re`:

```
import re

string = 'asd 21 0asd \asd1 1\ ds $a ds d3 \'112309a 2300 [qasd[ ]23 .asd///'

clean = re.sub(r'\W+', '_', string).strip('_')  # replace non-word chars with _, removing trailing

print(clean)
```

Reply:
@brisvag @Czaki while agree with the current regex, I don't want to change the capitalization of anything in the layer names. My use case is making segmentation masks for deep learning training data, and it's been useful to make layers that have almost the same filename as the images I am labeling, so I think we should mostly trust users here but agree about removing escaped chars, slashes, etc.

Reply:
I _think_ chars like `$` could be allowed in filenames on some filesystems see https://en.wikipedia.org/wiki/Filename#Comparison_of_filename_limitations for examples, but maybe there is a correct choice here that others are familiar with?

Reply:
I think that your use case require same ban list on every system. Otherwise your workflow may stop working only because change of OS. And adjust policy of setting labels name to not break such rules.

Reply:
@AndrewAnnex are you suggesting to leave spaces, brackets, other special signs, and only remove specific things? Can you list those?

I generally lean towards what @Czaki was saying, better have a general rule; and I think we could save a lot of non-tech-savvy users *lots* of headaches if we don't autogenerate filenames with spaces in it.

Reply:
@AndrewAnnex I also think we do have to sanitize the name in order to avoid issues with different file systems on different os. Would you still be interested in getting this PR over the line? Otherwise we could finish it as well.

Reply:
At the PR party we experimented with saving names with forbidden characters, and amazingly, it Just Worked — the (macOS) native file dialog just (silently) replaced `/` with `:`.

> I think we could save a lot of non-tech-savvy users lots of headaches if we don't autogenerate filenames with spaces in it.

I think this is misguided. Filenames with spaces are allowed by all OSs and completely common among scientists, and I don't think it's the place of the napari save dialog to teach them about good filename habits. (That is a job for [this excellent normconf talk](https://www.youtube.com/watch?v=ES1LTlnpLMk).)

It is worth sanitising *actually* forbidden characters, but nothing else, in my opinion. There's a [whole library](https://github.com/thombashi/pathvalidate) for this, but I don't think we should pick up a whole new dependency. I suggest just using their regexes [here](https://github.com/thombashi/pathvalidate/blob/e0ace30aec8df6ac7490f3d1bf4dd7af16a5ee50/pathvalidate/_filename.py#L22) and [here](https://github.com/thombashi/pathvalidate/blob/e0ace30aec8df6ac7490f3d1bf4dd7af16a5ee50/pathvalidate/_base.py#L18-L23).

Reply:
yeah sorry for being silent on this issue, started a new position and I've been busy, also too busy to work on this for the next month or more anyways. I am for minimal modification as philosophically, if you can load a filename you should be able to save with the same filename. So if there are spaces, some strange characters they should be allowed.

Reply:
No worries, we'll wrap it up for you then once we have figured out what to do, but I pretty much agree with that philosophy. (But do note that users can manually edit layer names, so it's not always the case that the layer name corresponds to the loaded file name.)

SETI hey??? That is very cool! If you use napari to visualise radio signals please send screenshots! 😃

Reply:
(and tbh please send screenshots/movies/scripts if you have them of using napari at your previous position — we're mostly a bioimaging bunch but are working to make sure napari is general-purpose, so I personally get very excited when folks from different scientific domains drop by the repo. 😊)

Reply:
@jni yeah if a user changes a layer name no need to worry about updating/keeping it in sync with the filename corresponding to it. 

Actually I'm a software engineer/planetary scientist so my work using Napari was mostly optical remote sensing of planetary surfaces, so not into the radio signal stuff. But there is a lot of overlap with biomedical imaging and remote sensing/image processing for geology/photometry. 

Here's an example from my AGU talk on using one of the segment anything plugins with Mars 2020 rover imagery but I cropped out the napari ui unfortunately 
![example_sam_float2](https://github.com/napari/napari/assets/2126916/ea6e8ee7-1f13-4781-be10-da7b6c84ebad)

Reply:
> Here's an example from my AGU talk on using one of the segment anything plugins with Mars 2020 rover imagery

😍😍😍😍😍😍😍

> but I cropped out the napari ui unfortunately

wth! 😜😂

More seriously — I take it there is no demo code/paper/other public artifact of this work that we could point to / build on? But either way thanks for sharing, made my day! 😊

Reply:
> but I cropped out the napari ui unfortunately

On further reflection it's ok — I recognise the napari labels colors. 😃 #trademark